### ISSUE TITLE
Consistency with sorting: torch.RNN default vs torchtext.Iterator default

### ISSUE THREAD

**Issue by Deepblue129** [pytorch/text#70]
**Torch RNN Default:**
Using torch.RNN with batches it is required that the batch is sorted with decreasing lengths.

**Default Behavior:***
The default behavior of BucketIterator does not work well with Torch RNN.
```
    train_iter, dev_iter, test_iter = data.BucketIterator.splits(
        (train, dev, test),
        batch_sizes=(32, 256, 256),
        sort_key=lambda x: data.interleave_keys(-len(x.input), -len(x.output)),
        device=-1)  # Use CPU
```

By default `train_iter` is compatible with Torch RNN. All the batches are sorted by decreasing lengths.

But by default `dev_iter` and `test_iter` are not sorted by decreasing lengths. This is because `train=False` then `self.sort=True`. Then because of this issue, https://github.com/pytorch/text/issues/69 `dev_iter` and `test_iter` are actually shuffled.

**Possible Solution:**
It's possible that solving this issue https://github.com/pytorch/text/issues/69 will solve this one as well but I think these are two separate bugs.

**Comment by Deepblue129** [comment]
@nelson-liu Not sure we're on the same page.

Do not think I am asking for: "torchtext to provide an API for unshuffling your batch predictions"
Not asking for this either "sort the batches in my code".

The sequences batched together need to be sorted by decreasing lengths. If I have a batch with the sequences.

"1 2 3"
"1 2 3 4 5"
"1 2 3 4"

Torch RNN with batching and `pack_padded_sequence` require the sequences to be sorted in decreasing lengths as such:

"1 2 3 4 5"
"1 2 3 4"
"1 2 3"

For both inference and training.

**Comment by jekbradbury** [comment]
It's always possible for the user to provide a `sort_key` that's the negative of what they would otherwise provide. Would that be enough to make the default behavior consistent with the RNN API? Should we default to treating `sort_key`s as implicitly negative? Or provide an additional kwarg for `Iterator`s to reverse their sort order?

**Comment by Deepblue129** [comment]
@jekbradbury I think the sort_key should remain as it is. Otherwise, it would not be consistent with default sorting behavior. 

Maybe add a parameter `reverse` and have it defaulted to True. That would be consistent with sort and torch.RNN. Like you mentioned!

Going to close this issue. Unable to replicate my original bug.
